pkg/runtime.v2/framework/plugins/jobset/jobset.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jobset
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	jobsetv1alpha2 "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+	runtime "github.com/kubeflow/training-operator/pkg/runtime.v2"
+	"github.com/kubeflow/training-operator/pkg/runtime.v2/framework"
+)
+
+type JobSet struct {
+	client     client.Client
+	restMapper meta.RESTMapper
+	scheme     *apiruntime.Scheme
+}
+
+var _ framework.WatchExtensionPlugin = (*JobSet)(nil)
+var _ framework.ComponentBuilderPlugin = (*JobSet)(nil)
+
+const Name = "JobSet"
+
+func New(_ context.Context, c client.Client, _ client.FieldIndexer) (framework.Plugin, error) {
+	return &JobSet{
+		client:     c,
+		restMapper: c.RESTMapper(),
+		scheme:     c.Scheme(),
+	}, nil
+}
+
+func (j *JobSet) Name() string {
+	return Name
+}
+
+func (j *JobSet) Build(ctx context.Context, info *runtime.Info, trainJob *kubeflowv2.TrainJob) (client.Object, error) {
+	if info == nil || info.Obj == nil || trainJob == nil {
+		return nil, fmt.Errorf("runtime info or object is missing")
+	}
+	raw, ok := info.Obj.(*jobsetv1alpha2.JobSet)
+	if !ok {
+		return nil, nil
+	}
```

Comment:
Do we need to have it if this will be part of validation ?

Reply:
We need to implement these validations here since the validations are called before main phase (https://docs.google.com/presentation/d/1HyEsBa7hxWpIoBXaX6uECiB48FWB85SG1kx15mO8hug/edit#slide=id.g30596bfee76_0_0). The `info` is constructing during the main phase. Hense, the validator does not know the `info`.

Reply:
But how the Info object can be constructed with type different from JobSet?

Reply:
The info construction is performed in the Runtime. So, info constructed does not need to know the JobSet.
Actually, we do not need to pass the JobSet: https://github.com/kubeflow/training-operator/blob/22da8af373f18e8e51c1e466a35a7738340b8c7d/pkg/runtime.v2/runtime.go#L106